src/state/queries/post.ts

```diff
@@ -0,0 +1,236 @@
+import {AppBskyFeedDefs} from '@atproto/api'
+import {
+  useQuery,
+  useQueryClient,
+  useMutation,
+  QueryClient,
+} from '@tanstack/react-query'
+import {useSession} from '../session'
+
+export const RQKEY = (postUri: string) => ['post', postUri]
+
+export function getCachedPost(
+  queryClient: QueryClient,
+  uri: string,
+): AppBskyFeedDefs.PostView | undefined {
+  return queryClient.getQueryData(RQKEY(uri))
+}
+
+export function usePostQuery(uri: string | undefined) {
+  const {agent} = useSession()
+  return useQuery<AppBskyFeedDefs.PostView>(
+    RQKEY(uri || ''),
+    async () => {
+      const res = await agent.getPosts({uris: [uri!]})
+      if (res.success && res.data.posts[0]) {
+        return res.data.posts[0]
+      }
+
+      throw new Error('No data')
+    },
+    {
+      enabled: !!uri,
+    },
+  )
+}
+
+export function usePostLikeMutation() {
+  const {agent} = useSession()
+  const queryClient = useQueryClient()
+  return useMutation<
+    {uri: string}, // responds with the uri of the like
+    Error,
+    {uri: string; cid: string}, // the post's uri and cid
+    {likeCount: number}
+  >(post => agent.like(post.uri, post.cid), {
+    onMutate(variables) {
+      // optimistically update the post-cache
+      let likeCount = 0
+      updatePostCache(queryClient, variables.uri, post => {
+        likeCount = post.likeCount || 0
+        return {
+          ...post,
+          likeCount: (post.likeCount || 0) + 1,
+          viewer: {...(post.viewer || {}), like: 'pending'},
+        }
+      })
+      return {likeCount}
+    },
+    onSuccess(data, variables) {
```

Comment:
React Query v5 is out which simplifies the optimistic update workflow. We no longer have to manually update the cache every time: 

https://medium.com/@stojanovic.nemanja71/optimistic-updates-in-tanstack-query-v5-dfbcbb124113#:~:text=So%20instead%20of%20waiting%20for,won't%20notice%20any%20delay.

Reply:
This applies to other optimistic updates as well. Not adding the comment everywhere fyi.

Reply:
The link above is how we should do things in _most_ cases. In this case, the post shadow is synced across components and it's better to put in these handlers